### DISCUSSION TITLE
资源文件的变动如何触发重新编译

### DISCUSSION THREAD

**Discussion by q962** [xmake-io/xmake#5422] [Q&A]
我需要将一个资源文件嵌入到 c 中。

目前是在主target 中，利用 before_build 和 before_link 函数分别生成目标文件和链接静态文件

问题是，如果我的资源文件发生变动，xmake 无法检测到。

如何让 xmake 检测资源文件的变动？

还有如果我想创建一个新的 target 并覆盖其 on_build 和 on_link。

又如何解决上述的问题

**Comment by q962** [comment]
同理，rule 又该如何解决这个问题

**Comment by q962** [comment]
我有一个可能的解决办法，使用 rule 方案。

xmake 在编译时记录目标文件的的 mtime，下载编译是如果目标文件的 mtime 和源文件mtime 不匹配则触发编译

**Comment by star-hengxing** [comment]
https://github.com/xmake-io/xmake/blob/dev/xmake%2Frules%2Fc51%2Fxmake.lua#L34-L43

**Reply by q962** [reply]
我看了一下，没明白那个函数的逻辑。
经过测试，我修改了相关文件，并没有触发事件

在 [`depend.on_changed`](https://github.com/xmake-io/xmake/blob/dev/xmake/modules/core/project/depend.lua#L204) 内部会利用传递过去的文件组合创建一个临时文件，用于代表该文件组。对其获取 mtime 参与比对。

这个临时文件的路径包含**精确到日**的目录名称。这是否意味着，及时该功能有效，那么隔天是否因为匹配不到昨天的文件而失败，从而导致无端的重新编译？

我获取到该临时文件内容，里面只有相关的文件路径。结合代码，发现这里函数内部关于 mtime 的日期判断不合理。

该功能应该参考 makefile 的逻辑，将相关的文件在**build目录中创建一个空文件**，

我认为函数应该参考 makefile 做出以下修改，或者提供一个新的函数
```lua
function depend.on_changed(cb, files)

    local changed = {};

    for _, file in ipairs(files) do
        local fake_obj = buildir .. file .. ".fake_obj"
        if not isFile(fake_obj ) or mtime(file) > mtime(fake_obj ) then
            touch(fake_obj )
            changed.push(file)
        end
    end

    if #changed then
        cb(changed)
    end
end

depend.on_changed(function (changed_files)
     -- 这里可以对发生改变的文件做出处理，和 makefile 的逻辑一致
end, {
    files = {}
})
```

**Reply by star-hengxing** [reply]
不知道你怎么写的，xmake 内部很多地方都是走这个接口，应该是很稳定的。

@waruqi

**Reply by q962** [reply]
仔细研究了一下，是否是需要在函数内部创建传递过去的 files？
用法不是很明白。

**Reply by q962** [reply]
我将 sourcebatch.sourcefiles 中的文件增加 .a 后缀，将其结果作为 files 传递给 on_changed.
在回调中会生成所有 .a 文件

但是结果并不理想，只有删除文件会调用回调，修改则不会。

你给的例子中传递了 target:targetfile() 文件，xmake 是怎么知道哪些文件的修改会导致该文件的需要更新?

我模仿了一下，我将生成 c 文件的完整路径传递过去，并在内部生成该 c 文件。只会在一开的时候调用一次。之后，不论是修改sourcebatch.sourcefiles 中的文件，还是删除，都无法触发回调。

**Comment by q962** [comment]
整个逻辑非常简单，用 makefile 描述如下

```makefile

file.c: %.a
	生成 c 文件

%.a: %. data
	生成 .a 文件
```

和xmake现在做的事是一样的。只是需要控制生成的文件和需要链接的目标文件/组合

**Reply by star-hengxing** [reply]
那就是常见的 codegen 场景，xmake 里都有很多例子，比如 yacc rule

**Comment by q962** [comment]
或许可以尝试 set_kind("object")  然后修改  target:objectfiles()

**Reply by q962** [reply]
搞不懂，搞不懂

**Comment by waruqi** [comment]
类似 bin2c? 参考下 https://github.com/xmake-io/xmake/blob/dev/xmake/rules/utils/bin2c/xmake.lua

**Reply by q962** [reply]
**我都气笑了🤣🤣🤣**

参考你给的文件修改为 before_buildcmd_files 模式 ，能够进行检测了。但是会导致一个文件的修改全部重新生成。
然后我就在想，如果使用 before_buildcmd_file 是不是就可以只针对被修改的文件进行重新生成了。

然后戏剧性的一幕来了。需要生成的文件有 20 左右

在 before_buildcmd_files 模式下，会导致最后一个文件生成失败。
！很无语，单独拎出来调用可以，伪造最后一文件，这个文件又失败了，之前的最后一个文件却没问题了。

在 before_buildcmd_file 模式下，会导致了命令死循环。进程个数随机，相关文件随机。
只有占用cpu，内存无变动。等了5分钟左右没有任何进展。

**每条命令单独拎出来，一个都没问题。**

```lua
batchcmds:execv("sh", {
    "-c",
    vformat(table.concat({
        "cd ",
        "$(scriptdir)/res",
        "&& ",
        "objcopy",
        "-I", "binary",
        "-O", "elf64-x86-64",
        "-B", "i386:x86-64",
        filepath, "$(projectdir)/"..a_filepath
    }, " ")):gsub("\\","/")
})
```

**Reply by q962** [reply]
然后我将命令收集起来放到单独的 lua 文件中 使用 xmake lua 执行它。模拟 before_buildcmd_files 模式，不会出现问题。

**Comment by q962** [comment]
还是自己写函数处理吧，给我整的头都大了
